File: src/main/java/com/coalesce/permissions/RankManager.java
package com.coalesce.permissions;

import com.coalesce.Bot;
import com.coalesce.utils.Streams;
import lombok.Getter;
import net.dv8tion.jda.core.entities.Guild;
import net.dv8tion.jda.core.entities.Member;
import net.dv8tion.jda.core.entities.Role;

import java.util.*;

public class RankManager {
    private static RankManager instance;
    private final static Object lock = new Object();

    public static RankManager getInstance() {
        RankManager inst = instance;
        if (inst == null) {
            synchronized (lock) {
                inst = instance;
                if (inst == null) {
                    inst = new RankManager();
                    instance = inst;
                }
            }
        }
        return inst;
    }

    @SuppressWarnings("MismatchedQueryAndUpdateOfCollection") // TODO: Update the map with ranks from json files.
    private final @Getter Map<Role, Map<String, Boolean>> ranks = new HashMap<>();
    private final @Getter Set<WrappedUser> users = new HashSet<>();

    private RankManager() {
        Bot.getInstance().getJda().getGuilds().stream().map(Guild::getMembers).parallel().forEach(members -> members.stream().forEach(member -> users.add(new WrappedUser(member))));
    }

    public Map<String, Boolean> getPermissions(Member member) {
        Optional<WrappedUser> optional = users.stream().filter(it -> it.getMember().equals(member)).findFirst();
        if (!optional.isPresent()) {
            return new HashMap<>();
        }
        Map<String, Boolean> applicablePerms = new HashMap<>();
        WrappedUser user = optional.get();
        Streams.reverse(ranks.entrySet().stream()
                .filter(it -> member.getRoles().contains(it.getKey()))
                .sorted()).forEachOrdered(it -> applicablePerms.putAll(it.getValue()));
        applicablePerms.putAll(user.getPermissions());
        return applicablePerms;
    }
}

File: src/main/kotlin/com/coalesce/permissions/RankManager.kt
package com.coalesce.permissions

import com.coalesce.Bot
import com.coalesce.utils.reverse
import net.dv8tion.jda.core.entities.Guild
import net.dv8tion.jda.core.entities.Member
import net.dv8tion.jda.core.entities.Role
import java.util.*

class RankManager private constructor() {
    private val ranks = HashMap<Role, Map<String, Boolean>>()
    private val users = HashSet<WrappedUser>()

    init {
        Bot.instance.jda.guilds.stream().map(Guild::getMembers).parallel().forEach({ it.forEach { users.add(WrappedUser(it)) } })
    }

    fun getPermissions(member: Member): Map<String, Boolean> {
        val optional = users.stream().filter { it.member == member }.findFirst()
        if (!optional.isPresent) {
            return HashMap()
        }
        val applicablePerms = HashMap<String, Boolean>()
        val user = optional.get()
        ranks.entries.stream()
                .filter { it -> member.roles.contains(it.key) }
                .sorted().reverse().forEachOrdered { it -> applicablePerms.putAll(it.value) }
        applicablePerms.putAll(user.permissions)
        return applicablePerms
    }

    fun hasPermission(member: Member, vararg permissions: String): Boolean {
        val permsMap = getPermissions(member)
        val perms = permsMap.entries.map { it.key }
        return Arrays.stream(permissions).allMatch { perms.contains(it) }
    }

    companion object {
        val instance = RankManager()
    }
}